### PR TITLE
PT-1933: Allow pushing phenotype details

### PR DESCRIPTION
**Pull request by veronikaslc**
Fixed.

### CODE REVIEW

**Comment by veronikaslc** [comment]
Also fixes [PT-2459: Selection of Clinical Symptoms and physical findings triggers pushing of Prenatal and perinatal phenotypes and vice versa](https://phenotips.atlassian.net/browse/PT-2459).

**Comment by veronikaslc** [comment]
@allasm @meatcar please, review

**Comment by meatcar** [comment]
:+1:

File: components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java [hunk 1]
```diff
@@ -394,17 +395,66 @@ private void updateFeaturesFromJSON(XWikiDocument doc, BaseObject data, XWikiCon
...
+        doc.removeXObjects(FeatureMetadatum.CLASS_REFERENCE);
+        for (Feature feature : this.features) {
+            @SuppressWarnings("unchecked")
+            Map<String, FeatureMetadatum> metadataMap = (Map<String, FeatureMetadatum>) feature.getMetadata();
+            if (metadataMap.isEmpty() && feature.getNotes().isEmpty()) {
+                continue;
+            }
+
+            BaseObject metaObject = doc.newXObject(FeatureMetadatum.CLASS_REFERENCE, context);
+            metaObject.set(PhenoTipsFeature.META_PROPERTY_NAME, ((PhenoTipsFeature) feature).getPropertyName(),
```
Comment (meatcar) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: Why instantiate `feature` as `PhenoTipsFeature` in the beginning of the for-loop? Saves you this cast here, and everything else would still work the same. Likewise for the `updateCategories` method.

Comment (veronikaslc) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: Not sure about the time saving here.. Won't the cast happen anyways?

Comment (meatcar) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: Its not a time saving, it'll just get rid of the `(PhenoTipsFeature)` all over the place. A bit cleaner, thats all.

Comment (allasm) [inline components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java]: I'm re-reading the entire pull request right now, but I think whatever makes PhenoTipsFeature different from Feature should just be in Feature to start with. If a PhenoTipsFeature may have "properties", a generic Feature should as well

**Comment by veronikaslc** [comment]
Updated. @allasm please re-review.

File: components/patient-data/api/src/main/java/org/phenotips/data/Feature.java [hunk 2]
```diff
@@ -91,4 +92,19 @@
...
+    /**
+     * Returns a feature property name. Constructed from type property prefixing the {@code negative_} prefix for
+     * {@link #isPresent() negative observations}.
+     *
+     * @return the feature property name, usually a string ending in {@code phenotype}
+     */
+    String getPropertyName();
+
+    /**
+     * Returns optional custom categories for this non-standard feature.
```
Comment (allasm) [inline components/patient-data/api/src/main/java/org/phenotips/data/Feature.java]: should document what is returned for standard features

File: components/patient-data/api/src/main/java/org/phenotips/data/Feature.java [hunk 3]
```diff
@@ -91,4 +92,19 @@
      * @todo move to VocabularyProperty and/or implement a VocabularyProperty-to-PhenotipsPropertyName mapping service
      */
     String getValue();
+
+    /**
+     * Returns a feature property name. Constructed from type property prefixing the {@code negative_} prefix for
```
Comment (allasm) [inline components/patient-data/api/src/main/java/org/phenotips/data/Feature.java]: I would also document WHAT is property name (e.g. is it "property name as stored in PhenoTips"?) property is a very generic term

**Comment by veronikaslc** [comment]
Code is rebased into push-1.3m2 branch.

**Pull request closed without merging**